gateway/src/cron.rs

```diff
@@ -0,0 +1,427 @@
+use ipc_sdk::ValidatorSet;
+use primitives::{TCid, THamt};
+use std::cmp::Ordering;
+use std::collections::HashSet;
+
+pub type HashOutput = Vec<u8>;
+const RATIO_NUMERATOR: u16 = 2;
+const RATIO_DENOMINATOR: u16 = 3;
+
+/// Checkpoints propagated from parent to child to signal the "final view" of the parent chain
+/// from the different validators in the subnet.
+#[derive(Clone, Debug, Serialize_tuple, Deserialize_tuple, PartialEq, Eq)]
+pub struct CronCheckpoint {
+    pub epoch: ChainEpoch,
+    pub validators: ValidatorSet,
+    pub top_down_msgs: Vec<StorableMsg>,
+}
+
+impl CronCheckpoint {
+    /// Hash the checkpoint.
+    ///
+    /// To compare the cron checkpoint and ensure they are the same, we need to make sure the
+    /// validators and top_down_msgs are the same. However, the top_down_msgs and validators are vec,
+    /// they may contain the same content, but their orders are different. In this case, we need to
+    /// ensure the same order is maintained in the cron checkpoint submission.
+    ///
+    /// To ensure we have the same consistent output for different submissions, we require:
+    ///     - validators are sorted by `net_addr` in string ascending order
+    ///     - top down messages are sorted by (from, to, nonce) in descending order
+    ///
+    /// Actor will not perform sorting to save gas. Client should do it, actor just check.
+    fn hash(&self) -> anyhow::Result<HashOutput> {
+        // check validators
+        let validators = self.validators.validators();
+        for i in 1..validators.len() {
+            match validators[i - 1].net_addr.cmp(&validators[i].net_addr) {
+                Ordering::Less => {}
+                Ordering::Equal => return Err(anyhow!("validators not unique")),
+                Ordering::Greater => return Err(anyhow!("validators not sorted")),
+            };
+        }
+
+        // check top down msgs
+        for i in 1..self.top_down_msgs.len() {
+            match compare_top_down_msg(&self.top_down_msgs[i - 1], &self.top_down_msgs[i])? {
+                Ordering::Less => {}
+                Ordering::Equal => return Err(anyhow!("top down messages not distinct")),
+                Ordering::Greater => return Err(anyhow!("top down messages not sorted")),
+            };
+        }
+
+        let mh_code = Code::Blake2b256;
+        // TODO: to avoid serialization again, maybe we should perform deserialization in the actor
+        // TODO: dispatch call to save gas? The actor dispatching contains the raw serialized data,
+        // TODO: which we dont have to serialize here again
+        Ok(mh_code.digest(&to_vec(self).unwrap()).to_bytes())
+    }
+}
+
+/// Track all the cron checkpoint submissions of an epoch
+#[derive(Serialize_tuple, Deserialize_tuple, PartialEq, Eq, Clone)]
+pub struct CronSubmission {
+    /// All the submitters
+    submitters: TCid<THamt<Address, ()>>,
+    /// The most submitted hash. Using set because there might be a tie
```

Comment:
I think this structure is too complex, right? You only need to keep track of the checkpoints submitted and their count. I don't think we need HAMTs (at least for now) in any of these. Mir can't support over a hundred validators, so I don't think we will have more than a 66 validators here. Every time we get a hamt we pay some gas to access the datastore and pass the FVM boundary, so I would simplify this a bit.

What about something like: 
```
submissions: HashMap<HashOutput, CronCheckpoint>
submitters: HashMap<HashOutput, Vec<Address>>
``` 
You cache all the checkpoints submitted, and use index through hash to check the submitters and the count.

Reply:
as discussed over slack call, we will keep the current for now. Later we can do a comparison over which method will have a smaller gas.